projects/rocr-runtime/runtime/hsa-runtime/core/inc/counted_queue_manager.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright © Advanced Micro Devices, Inc., or its affiliates. 
+ * 
+ * SPDX-License-Identifier: MIT
+ */
+
+#ifndef HSA_RUNTME_CORE_INC_COUNTED_QUEUE_MANAGER_H_
+#define HSA_RUNTME_CORE_INC_COUNTED_QUEUE_MANAGER_H_
+
+
+#include "hsa.h"
+#include "hsa_ext_amd.h"
+#include "core/inc/agent.h"
+#include "core/inc/queue.h"
+#include "core/inc/runtime.h"
+
+#include "map"
+
+namespace rocr {
+namespace core {
+
+// Wrapper for real hardware queue tracking and its references
+// Each entry in queue pool is of this type
+struct HardwareQueue {
+  hsa_queue_t* hw_queue;  // Actual hardware queue
+  uint32_t use_count;     // Internal reference count
+  hsa_amd_queue_priority_t priority;
+  hsa_agent_t agent;
+
+  HardwareQueue(hsa_queue_t* q, hsa_amd_queue_priority_t prio, hsa_agent_t ag)
+      : hw_queue(q), use_count(0), priority(prio), agent(ag) {}
+};
+
+// Wrapper around HW queue to provide unique logical handles to multiple users, even when same HW
+// queue is used. Also store callbacks per logical handle, not HW handle
+struct CountedQueue {
+  HardwareQueue* hw_queue;  // Pointer to shared hardware queue
+  // callback per unique logical handle, not hardware handle
+  void (*callback)(hsa_status_t, hsa_queue_t*, void*);
+  void* callback_data;
+
+  CountedQueue(HardwareQueue* hw, void (*cb)(hsa_status_t, hsa_queue_t*, void*), void* data)
+      : hw_queue(hw), callback(cb), callback_data(data) {}
+};
+
+// Singleton manager for a pool of counted queues
+class CountedQueuePoolManager {
```

Comment:
I would make the CountedQueuePoolManager a member of the GpuAgent class and initialize it in the GpuAgent constructor. That way we do not have to make it a singleton and check for IsInstanceCreated() each time.

Then inside  hsa_amd_counted_queue_acquire implementation, we would call:

status = agent->CountedQueueCreate(....);

Which will invoke the CountedQueuePoolManager for the GpuAgent.

For CPU agents, hsa_amd_counted_queue_acquire can return error. 
For AIE agents, you can leave these functions empty for now. We can figure out how to implement it as we closer to finalizing this patch.